File: ios/ParkPass/Screens/VisitStamp.swift
import SwiftUI

/// The rubber-stamp mark for a visited park, drawn on the web version's
/// 120-unit grid (VisitStamp.tsx). Rotation, opacity and the press animation
/// are applied by the caller.
struct VisitStamp: View {
    let iso: String
    @Environment(L10n.self) private var l10n

    var body: some View {
        let date = l10n.stampDate(iso)
        let visited = l10n.t("detail.stampVisited")

        Canvas { context, size in
            let k = min(size.width, size.height) / 120
            let ink = Color.sage700
            let center = CGPoint(x: 60 * k, y: 60 * k)

            context.stroke(Path(ellipseIn: CGRect(x: 4 * k, y: 4 * k, width: 112 * k, height: 112 * k)), with: .color(ink), lineWidth: 2.6 * k)
            context.stroke(Path(ellipseIn: CGRect(x: 16 * k, y: 16 * k, width: 88 * k, height: 88 * k)), with: .color(ink), lineWidth: 1 * k)

            // "PARKPASS" along the top arc (r = 48), centred on the top.
            let arcFont = Font.figtree(9.5 * k, .bold)
            let tracking = 2.6 * k
            let glyphs = "PARKPASS".map { ch in
                context.resolve(Text(String(ch)).font(arcFont).foregroundColor(ink))
            }
            let widths = glyphs.map { $0.measure(in: size).width }
            let total = widths.reduce(0, +) + tracking * Double(glyphs.count - 1)
            let radius = 48 * k
            var angle = -Double.pi / 2 - (total / radius) / 2
            for (glyph, width) in zip(glyphs, widths) {
                let mid = angle + (width / 2) / radius
                var g = context
                g.translateBy(x: center.x + radius * cos(mid), y: center.y + radius * sin(mid))
                g.rotate(by: .radians(mid + .pi / 2))
                g.draw(glyph, at: .zero, anchor: .bottom)
                angle += (width + tracking) / radius
            }

            func centred(_ text: String, y: Double, font: Font, tracking: Double) {
                let resolved = context.resolve(Text(text).font(font).tracking(tracking).foregroundColor(ink))
                context.draw(resolved, at: CGPoint(x: center.x, y: y * k), anchor: .bottom)
            }
            centred(visited, y: 55, font: .figtree(13 * k, .heavy), tracking: 2.2 * k)
            centred(date, y: 71, font: .figtree(10.5 * k, .bold), tracking: 0.8 * k)
            centred("★★★", y: 86, font: .system(size: 9 * k), tracking: 3 * k)
        }
        .accessibilityElement()
        .accessibilityLabel(l10n.t("detail.stampAlt", ["date": date]))
    }
}

File: android/app/src/main/java/com/hopchenko/parkpass/ui/detail/VisitStamp.kt
package com.hopchenko.parkpass.ui.detail

import android.graphics.Paint
import android.graphics.Path
import android.graphics.RectF
import android.graphics.Typeface
import android.os.Build
import androidx.compose.foundation.Canvas
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.drawIntoCanvas
import androidx.compose.ui.graphics.nativeCanvas
import androidx.compose.ui.graphics.toArgb
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.core.content.res.ResourcesCompat
import com.hopchenko.parkpass.R
import com.hopchenko.parkpass.core.Lang
import com.hopchenko.parkpass.data.formatStampDate
import com.hopchenko.parkpass.ui.theme.ParkPassColors

/**
 * The rubber-stamp mark for a visited park, drawn on the web version's
 * 120-unit grid (VisitStamp.tsx). Rotation, opacity and the press animation
 * are applied by the caller.
 */
@Composable
fun VisitStamp(iso: String, lang: Lang, modifier: Modifier = Modifier) {
    val context = LocalContext.current
    val visited = stringResource(R.string.detail_stampVisited)
    val date = formatStampDate(iso, lang)
    val label = stringResource(R.string.detail_stampAlt, date)
    val ink = ParkPassColors.Sage700

    val base = remember { ResourcesCompat.getFont(context, R.font.figtree) ?: Typeface.DEFAULT }
    fun weighted(weight: Int): Typeface =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) Typeface.create(base, weight, false)
        else Typeface.create(base, if (weight >= 600) Typeface.BOLD else Typeface.NORMAL)
    val bold = remember(base) { weighted(700) }
    val extraBold = remember(base) { weighted(800) }

    Canvas(modifier.semantics { contentDescription = label }) {
        val k = size.minDimension / 120f
        drawCircle(ink, radius = 56f * k, style = Stroke(width = 2.6f * k))
        drawCircle(ink, radius = 44f * k, style = Stroke(width = 1f * k))

        drawIntoCanvas { canvas ->
            val c = canvas.nativeCanvas
            fun paint(typeface: Typeface, sizeUnits: Float, spacingUnits: Float) = Paint(Paint.ANTI_ALIAS_FLAG).apply {
                color = ink.toArgb()
                this.typeface = typeface
                textSize = sizeUnits * k
                // SVG letter-spacing is absolute; Paint wants ems.
                letterSpacing = spacingUnits / sizeUnits
            }

            // "PARKPASS" along the top arc: M 12,60 A 48,48 0 0 1 108,60.
            val arcPaint = paint(bold, 9.5f, 2.6f)
            val arc = Path().apply {
                addArc(RectF(12f * k, 12f * k, 108f * k, 108f * k), 180f, 180f)
            }
            val arcLength = (Math.PI * 48f * k).toFloat()
            val word = "PARKPASS"
            c.drawTextOnPath(word, arc, (arcLength - arcPaint.measureText(word)) / 2f, 0f, arcPaint)

            fun centred(text: String, y: Float, p: Paint) {
                c.drawText(text, 60f * k - p.measureText(text) / 2f, y * k, p)
            }
            centred(visited, 55f, paint(extraBold, 13f, 2.2f))
            centred(date, 71f, paint(bold, 10.5f, 0.8f))
            centred("★★★", 86f, paint(base, 9f, 3f))
        }
    }
}
